services/api_gateway/src/main/java/com/futurewei/alcor/apigateway/utils/KeystoneClient.java

```diff
@@ -0,0 +1,333 @@
+                return;
+            }
+            HttpHeaders headers = new HttpHeaders();
+            headers.setContentType(MediaType.APPLICATION_JSON);
+            headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
+            HttpEntity<String> entity = new HttpEntity<>(buildLocalTokenParams(), headers);
+            HttpEntity<String> response = restTemplate.postForEntity(baseUrl + TOKEN_URL + "?nocatalog", entity, String.class);
+
+            localToken = response.getHeaders().getFirst(VALIDATE_TOKEN_HEADER);
+            JsonNode result = json2Map(response.getBody());
+            JsonNode token = result.path("token");
+            String expireDateStr = token.path("expires_at").asText();
+            if (!"null".equals(expireDateStr)) {
+                expireDateStr = expireDateStr.replace("000Z", "+0000");
+                try {
+                    localTokenExpireDate = dateFormat.parse(expireDateStr);
+                } catch (ParseException e) {
+                    LOG.error("Get Alcor Token failed, {}", e.getMessage());
+                    localToken = "";
+                }
+            }
+        }
+
+    }
+
+    public String verifyToken(String token){
+        try {
+
+            TokenEntity tokenEntity = cache.get(token);
+            if(tokenEntity != null){
+                return tokenEntity.isExpired() ? "" : tokenEntity.getProjectId();
+            }
+
+            checkEndPoints();
+            getLocalToken();
+
+            HttpHeaders headers = new HttpHeaders();
+            headers.setContentType(MediaType.APPLICATION_JSON);
+            headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
+            headers.set(VALIDATE_TOKEN_HEADER, token);
+            headers.set(AUTH_TOKEN_HEADER, localToken);
+
+            HttpEntity<String> entity = new HttpEntity<>(null, headers);
+            ResponseEntity<String> response = restTemplate.exchange(baseUrl + TOKEN_URL + "?nocatalog",
+                    HttpMethod.GET, entity, String.class);
+
+            // check headers
+            if(response.getStatusCode().equals(HttpStatus.OK)){
+
+
+                String resultStr = response.getBody();
+                JsonNode result = json2Map(resultStr);
+                JsonNode tokenNode = result.path("token");
+
+                TokenEntity te = new TokenEntity(token,false);
+                JsonNode user = tokenNode.path("user");
+                te.setUser(user.path("name").asText(""));
+                te.setUserId(user.path("id").asText(""));
+
+                String expireDateStr = tokenNode.path("expires_at").asText();
+                expireDateStr = expireDateStr.replace("000Z", "+0000");
+                Date expireDate = dateFormat.parse(expireDateStr);
+                te.setExpireAt(expireDate);
+
+                if(tokenNode.has("roles")){
```

Comment:
Can we give an OpenStack KeyStone link about different roles to improve the readability?